pyinv/assets/models/asset.py

```diff
@@ -41,3 +44,30 @@ def asset_codes(self) -> List[str]:
 
     def __str__(self) -> str:
         return f"{self.display_name} ({self.first_asset_code})"
+
+    def add_asset_code(self, code_type: AssetCodeType, code: Optional[str]) -> AssetCode:
+        """
+        Add an asset code to an asset.
+
+        :raises ValueError: Unable to generate code, or unrecognised code type.
+        :raises django.db.IntegrityError: The specifed code already exists
+        """
+        def _generate_code(strategy: AssetCodeStrategy) -> Optional[str]:
+            code = strategy.generate_new_code()
+
+            # If we have generated a code, check it's not already in use.
+            if code is not None and AssetCode.objects.filter(code=code).exists():
+                # Try again.
+                return _generate_code(strategy)  # pragma: nocover
+
+            return code
+
+        if code is None:
+            code = _generate_code(AssetCodeType.get_strategy(code_type))
+
+            if code is None:
+                raise ValueError(
```

Comment:
Question: When will this case ever get hit? If it can't automatically generate one, then it probably means there are no valid ones left, in which case adding one yourself probably won't work.

Reply:
Not all asset code types support generation intentionally. The generation of SR codes is very coupled to the git based system and would be a huge pain to implement otherwise. Similar problems arise with arbitrary strings.